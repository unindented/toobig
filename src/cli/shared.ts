import { cosmiconfigSync } from "cosmiconfig";
import { existsSync, lstatSync } from "fs-extra";

// eslint-disable-next-line @typescript-eslint/no-var-requires
const pkg = require("../../package.json") as { name: string };

export const parserConfiguration = {
  "boolean-negation": true,
  "camel-case-expansion": false,
  "combine-arrays": false,
  "dot-notation": false,
  "duplicate-arguments-array": true,
  "flatten-duplicate-arrays": false,
  "halt-at-non-option": false,
  "negation-prefix": "no-",
  "parse-numbers": false,
  "populate--": false,
  "set-placeholder-key": false,
  "short-option-groups": false,
  "strip-aliased": false,
  "strip-dashed": false,
  "unknown-options-as-args": false,
};

export const configParser = (configPath?: string): object => {
  const explorer = cosmiconfigSync(pkg.name);
  const result = isFile(configPath)
    ? explorer.load(configPath)
    : explorer.search(configPath);

  return result != null && typeof result.config === "object"
    ? (result.config as object)
    : {};
};

const isFile = (path?: string): path is string =>
  !!(path && existsSync(path) && lstatSync(path).isFile());
