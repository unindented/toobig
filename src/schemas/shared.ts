import Joi from "@hapi/joi";

export const reportersConfigSchema = Joi.array().items(
  Joi.alternatives().try(
    Joi.string(),
    Joi.array().items(Joi.string().required(), Joi.object().required())
  )
);

export const resultsConfigSchema = Joi.array().items(
  Joi.object({
    path: Joi.string().required(),
    size: Joi.number().required(),
    maxSize: Joi.number().required(),
  }).required()
);
