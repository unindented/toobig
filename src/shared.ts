import { parse } from "url";

import axios from "axios";
import { Instance as ChalkInstance, supportsColor } from "chalk";
import { createReadStream, createWriteStream, ensureFileSync } from "fs-extra";

import { InputStream, OutputContext, OutputStream, Result } from "./types";

export { supportsColor };

export const isOverBudget = ({ size, maxSize }: Result): boolean =>
  size >= maxSize;

export const getTotalSize = (results: Result[]): number =>
  results.reduce((acc, { size }) => acc + size, 0);

export const getOutputContext = (supportsColor: boolean): OutputContext => {
  return {
    colors: new ChalkInstance(!supportsColor ? { level: 0 } : undefined),
  };
};

export const getOutputStream = (
  output: string | NodeJS.WritableStream
): OutputStream => {
  if (output === "stdout" || output === "stderr") {
    return process[output];
  }

  if (typeof output === "string") {
    ensureFileSync(output);
    return createWriteStream(output);
  }

  return output;
};

export const endOutputStream = (stream: OutputStream): void => {
  if (stream === process.stdout || stream === process.stderr) {
    return;
  }

  stream.end();
};

export const getInputStream = async (
  input: string | NodeJS.ReadableStream
): Promise<InputStream> => {
  if (input === "stdin") {
    return process.stdin;
  }

  if (typeof input === "string") {
    const { protocol } = parse(input);
    return protocol == null
      ? createReadStream(input)
      : await createAxiosStream(input);
  }

  return input;
};

const createAxiosStream = async (
  input: string
): Promise<NodeJS.ReadableStream> => {
  const { data: stream } = await axios.get<NodeJS.ReadableStream>(input, {
    responseType: "stream",
    timeout: 10000,
  });
  return stream;
};
