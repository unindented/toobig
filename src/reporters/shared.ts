import { basename, dirname, relative, resolve, sep } from "path";

import bytes from "bytes";
import terminalLink from "terminal-link";

import { getTotalSize, isOverBudget, supportsColor } from "../shared";
import { OutputContext, Result } from "../types";

export { getTotalSize, isOverBudget, supportsColor };

export const formatSize = ({ size }: Result): string => bytes(size);

export const formatMaxSize = ({ maxSize }: Result): string => bytes(maxSize);

export const formatSizeVsMaxSize = (
  result: Result,
  { colors, colorOverride }: OutputContext
): string => {
  const overMaxSize = isOverBudget(result);
  const color = colorOverride
    ? colorOverride
    : overMaxSize
    ? colors.bold.red
    : colors.dim;
  const operator = overMaxSize ? ">=" : "<";

  return color(`${formatSize(result)} ${operator} ${formatMaxSize(result)}`);
};

export const formatTotalSize = (results: Result[]): string =>
  bytes(getTotalSize(results));

export const formatPath = (
  { path }: Result,
  { colors }: OutputContext
): string => {
  const relativePath = relative(process.cwd(), path);
  const dir = dirname(relativePath);
  const base = basename(relativePath);
  return `${dir !== "." ? colors.dim(dir + sep) : ""}${colors.bold(base)}`;
};

export const linkPath = (result: Result, options: OutputContext): string => {
  const formattedPath = formatPath(result, options);
  return terminalLink(
    formattedPath,
    `file://${resolve(process.cwd(), result.path)}`,
    {
      fallback: () => formattedPath,
    }
  );
};
