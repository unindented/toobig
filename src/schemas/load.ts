import Joi from "@hapi/joi";

import { reportersConfigSchema, resultsConfigSchema } from "./shared";

export const loadConfigSchema = Joi.object({
  input: Joi.alternatives().try(Joi.string(), resultsConfigSchema).required(),
  reporters: reportersConfigSchema,
}).unknown();
