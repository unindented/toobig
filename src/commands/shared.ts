import { CompositeReporter } from "../reporters";
import {
  filterResults,
  getInputStream,
  isOverBaseline,
  isOverBudget,
  isUnderBaseline,
  readInputStream,
} from "../shared";
import {
  Reporter,
  ReporterConfig,
  ReporterConstructor,
  Results,
  ReturnValues,
} from "../types";

export const getCompositeReporter = (
  configReporters: readonly ReporterConfig[]
): Reporter => {
  const reporters = configReporters.map((configReporter) => {
    let reporter: Reporter;

    if (typeof configReporter === "string") {
      const ReporterClass = resolveReporter(configReporter);
      reporter = new ReporterClass();
    } else {
      const ReporterClass = resolveReporter(configReporter[0]);
      reporter = new ReporterClass(configReporter[1]);
    }

    return reporter;
  });

  return new CompositeReporter({ reporters });
};

const resolveReporter = (reporterName: string): ReporterConstructor => {
  let module: { default: ReporterConstructor };

  try {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    module = require(`../reporters/${reporterName}`);
  } catch (err) {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    module = require(reporterName);
  }

  return module.default;
};

export const loadResults = async ({
  results,
}: {
  results: string | Results;
}): Promise<Results> => {
  if (typeof results !== "string") {
    return results;
  }

  const stream = await getInputStream(results);
  const contents = await readInputStream(stream);
  return JSON.parse(contents) as Results;
};

export const reportResults = async ({
  results,
  baselines,
  reporter,
}: {
  results: Results;
  baselines?: Results;
  reporter: Reporter;
}): Promise<void> => {
  await reporter.onRunStart();

  for (const result of Object.values(results)) {
    await reporter.onResult(
      result,
      baselines ? baselines[result.path] : undefined
    );
  }

  await reporter.onRunComplete(results, baselines);
};

export const getReturnValues = ({
  results,
  baselines,
}: {
  results: Results;
  baselines?: Results;
}): ReturnValues => {
  const resultsUnderBaseline = filterResults(
    results,
    isUnderBaseline(baselines)
  );
  const resultsOverBaseline = filterResults(results, isOverBaseline(baselines));
  const resultsOverBudget = filterResults(results, isOverBudget);

  const anyUnderBaseline = Object.keys(resultsUnderBaseline).length > 0;
  const anyOverBaseline = Object.keys(resultsOverBaseline).length > 0;
  const anyOverBudget = Object.keys(resultsOverBudget).length > 0;

  return {
    results,
    anyUnderBaseline,
    anyOverBaseline,
    anyOverBudget,
  };
};
