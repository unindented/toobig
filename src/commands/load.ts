import { loadConfigSchema } from "../schemas";
import { getInputStream, isOverBudget } from "../shared";
import { LoadConfig, Result, ReturnValue } from "../types";

import { getCompositeReporter, reportResults } from "./shared";

export const loadAndReport = async (
  config: LoadConfig
): Promise<ReturnValue> => {
  await loadConfigSchema.validateAsync(config);

  const { input, reporters = ["default"] } = config;
  const compositeReporter = getCompositeReporter(reporters);

  const results = await loadResults({ input });
  await reportResults({ results, reporter: compositeReporter });

  const anyOverBudget = results.some(isOverBudget);
  return { results, anyOverBudget };
};

const loadResults = async ({
  input,
}: {
  input: string | Result[];
}): Promise<Result[]> => {
  if (typeof input !== "string") {
    return input;
  }

  const stream = await getInputStream(input);

  return new Promise((resolve, reject) => {
    const contents: string[] = [];

    stream.on("data", (data: string | Buffer) => {
      contents.push(data.toString());
    });

    stream.on("end", () => {
      resolve(JSON.parse(contents.join("")) as Result[]);
    });

    stream.on("error", (error: Error) => {
      reject(error);
    });
  });
};
