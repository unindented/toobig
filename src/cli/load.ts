import { Argv } from "yargs";

import { loadAndReport } from "../commands/load";
import { LoadConfig } from "../types";

import { configParser } from "./shared";

// eslint-disable-next-line @typescript-eslint/no-var-requires
const pkg = require("../../package.json") as { homepage: string };

export const command = "load [options]";

export const describe = "Load results from input file, and report them";

const isInputRequired = !!process.stdin.isTTY;
// istanbul ignore next
const inputDefault = !isInputRequired ? "stdin" : undefined;

export const builder = (yargs: Argv): Argv =>
  yargs
    .option("input", {
      describe: "Input file",
      type: "string",
      required: isInputRequired,
      default: inputDefault,
    })
    .option("reporters", {
      describe: "Reporters to process results",
      type: "array",
    })
    .option("config", {
      describe: "Config file, or directory to search for one",
      type: "string",
      default: ".",
      normalize: true,
      config: true,
      configParser,
    })
    .group(["input", "reporters", "config"], "Command options:")
    .example("$0 load results.json --reporters table summary", "")
    .example("$0 load --input results.json", "")
    .example("$0 load --input results.json --config myconfig.json", "")
    .example("$0 load --input results.json --reporters table summary", "")
    .example("$0 load < results.json", "")
    .example("$0 scan --reporters json | $0 load", "")
    .epilogue(`For more details, please see: ${pkg.homepage}`);

export const handler = async (argv: LoadConfig): Promise<boolean> => {
  const { anyOverBudget } = await loadAndReport(argv);
  return anyOverBudget;
};
