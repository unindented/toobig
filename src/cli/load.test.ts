jest.mock("../commands/load");

import { resolve } from "path";

import yargs, { Arguments } from "yargs";

import { loadAndReport } from "../commands/load";
import { LoadConfig } from "../types";

import * as loadCommand from "./load";
import { parserConfiguration } from "./shared";

// eslint-disable-next-line @typescript-eslint/no-var-requires
const pkg = require("../../package.json") as { name: string };

describe("load", () => {
  let parse: (arg: string) => Promise<string>;

  beforeEach(() => {
    (loadAndReport as jest.Mock).mockResolvedValue({
      results: [],
      anyOverBudget: false,
    });

    const parser = yargs
      .scriptName(pkg.name)
      .command<any>(loadCommand)
      .help()
      .parserConfiguration(parserConfiguration);

    parse = (arg: string): Promise<string> =>
      new Promise((resolve, reject) => {
        parser.parse(
          arg,
          (
            err: Error | undefined,
            _argv: Arguments<LoadConfig>,
            output: string
          ) => {
            if (err) {
              reject(err);
            } else {
              resolve(output);
            }
          }
        );
      });
  });

  describe("with `--help`", () => {
    let output: string;

    beforeEach(async () => {
      output = await parse("load --help");
    });

    it("shows help output", () => {
      expect(output).toMatchSnapshot();
    });
  });

  describe("with input", () => {
    beforeEach(async () => {
      await parse("load --input results.json");
    });

    it("calls `loadAndReport` with the right configuration", () => {
      expect(loadAndReport).toHaveBeenCalledWith(
        expect.objectContaining({
          input: "results.json",
        })
      );
    });
  });

  describe("with reporters", () => {
    beforeEach(async () => {
      await parse("load --input results.json --reporters table summary");
    });

    it("calls `loadAndReport` with the right configuration", () => {
      expect(loadAndReport).toHaveBeenCalledWith(
        expect.objectContaining({
          reporters: ["table", "summary"],
        })
      );
    });
  });

  describe("with a config file", () => {
    const configPath = resolve(
      __dirname,
      "../__fixtures__/load-config-valid.json"
    );
    const config = require(configPath);

    beforeEach(async () => {
      await parse(`load --config ${configPath}`);
    });

    it("calls `loadAndReport` with the right configuration", () => {
      expect(loadAndReport).toHaveBeenCalledWith(
        expect.objectContaining(config)
      );
    });
  });
});
