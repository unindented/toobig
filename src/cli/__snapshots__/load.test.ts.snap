// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`load with \`--help\` shows help output 1`] = `
"toobig load [options]

Load results from input file, and report them

Command options:
  --input      Input file                            [string] [default: \\"stdin\\"]
  --reporters  Reporters to process results                              [array]
  --config     Config file, or directory to search for one
                                                         [string] [default: \\".\\"]

Options:
  --version  Show version number                                       [boolean]
  --help     Show help                                                 [boolean]

Examples:
  toobig load results.json --reporters table summary
  toobig load --input results.json
  toobig load --input results.json --config myconfig.json
  toobig load --input results.json --reporters table summary
  toobig load < results.json
  toobig scan --reporters json | toobig load

For more details, please see: https://github.com/unindented/toobig"
`;
